app/views/contexts/grand_calculated_summary_context.py

```diff
@@ -0,0 +1,126 @@
+        # find any sections involved in the grand calculated summary (but only if they have started, to avoid evaluating the path if not necessary)
+        started_sections = [
+            key for key, _ in self._progress_store.started_section_keys(section_ids)
+        ]
+        routing_path_block_ids: list[str] = []
+
+        for section_id in started_sections:
+            if section_id == self.current_location.section_id:
+                routing_path_block_ids.extend(self.routing_path_block_ids)
+            else:
+                routing_path_block_ids.extend(
+                    # repeating calculated summaries are not supported at the moment, so no list item is needed
+                    self._router.routing_path(section_id).block_ids
+                )
+
+        return routing_path_block_ids
+
+    def build_groups_for_section(
+        self,
+        *,
+        section: Mapping,
+        return_to_block_id: str,
+        routing_path_block_ids: Iterable[str],
+    ) -> list[Mapping]:
+        return [
+            Group(
+                group_schema=group,
+                routing_path_block_ids=routing_path_block_ids,
+                answer_store=self._answer_store,
+                list_store=self._list_store,
+                metadata=self._metadata,
+                response_metadata=self._response_metadata,
+                schema=self._schema,
+                location=self.current_location,
+                language=self._language,
+                progress_store=self._progress_store,
+                return_to="grand-calculated-summary",
+                return_to_block_id=return_to_block_id,
+                summary_type="GrandCalculatedSummary",
+            ).serialize()
+            for group in section["groups"]
+        ]
+
+    def build_view_context(self) -> dict[str, dict]:
+        """
+        Build summary section with formatted total and change links for each calculated summary
+        """
+        # Type ignore: Block will exist at this point
+        block_id: str = self.current_location.block_id  # type: ignore
+        block: ImmutableDict = self._schema.get_block(block_id)  # type: ignore
+
+        calculation = block["calculation"]
+        calculated_summary_ids = get_calculation_block_ids_for_grand_calculated_summary(
+            block
+        )
+        routing_path_block_ids = self._blocks_on_routing_path(calculated_summary_ids)
+
+        calculated_section = self._build_grand_calculated_summary_section(block)
+
+        groups = self.build_groups_for_section(
+            section=calculated_section,
+            return_to_block_id=block_id,
+            routing_path_block_ids=routing_path_block_ids,
+        )
+        total = self._get_evaluated_total(
+            calculation=calculation["operation"],
+            routing_path_block_ids=routing_path_block_ids,
+        )
```

Comment:
This way you get one type ignore less, which is probably better, then extract id from created `block` (may fail linting):
```suggestion
        block: ImmutableDict = self._schema.get_block(self.current_location.block_id)  # type: ignore

        calculation = block["calculation"]
        calculated_summary_ids = get_calculation_block_ids_for_grand_calculated_summary(
            block
        )
        routing_path_block_ids = self._blocks_on_routing_path(calculated_summary_ids)

        calculated_section = self._build_grand_calculated_summary_section(block)

        groups = self.build_groups_for_section(
            section=calculated_section,
            return_to_block_id=block["id"],
            routing_path_block_ids=routing_path_block_ids,
        )
        total = self._get_evaluated_total(
            calculation=calculation["operation"],
            routing_path_block_ids=routing_path_block_ids,
        )
```

Reply:
I need the block id for line 105 as well so I've left this one be, but changed your other suggestion 👍

Reply:
if I use that though I'll need to type ignore twice instead of once?

Reply:
updated 👍